# Whisper with transformers

https://huggingface.co/openai/whisper-medium

https://github.com/huggingface/transformers/tree/main



In [3]:
from transformers import WhisperProcessor, WhisperForConditionalGeneration
from datasets import load_dataset

# load model and processor
processor = WhisperProcessor.from_pretrained("openai/whisper-medium")
model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-medium")
model.config.forced_decoder_ids = None

# load dummy dataset and read audio files
ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
sample = ds[0]["audio"]
input_features = processor(sample["array"], sampling_rate=sample["sampling_rate"], return_tensors="pt").input_features 

# generate token ids
predicted_ids = model.generate(input_features)
# decode token ids to text
transcription = processor.batch_decode(predicted_ids, skip_special_tokens=False)
print(transcription)

transcription = processor.batch_decode(predicted_ids, skip_special_tokens=True)
print(transcription)


C:\Users\tzurv\.conda\envs\whisper\Lib\site-packages\huggingface_hub\file_download.py:133: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\tzurv\.cache\huggingface\hub. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


Extracting data files:   0%|          | 0/1 [00:00<?, ?it/s]

Generating validation split: 0 examples [00:00, ? examples/s]

C:\Users\tzurv\.conda\envs\whisper\Lib\site-packages\transformers\generation\utils.py:1369: UserWarning: Using `max_length`'s default (448) to control the generation length. This behaviour is deprecated and will be removed from the config in v5 of Transformers -- we recommend using `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(


In [7]:
type(sample["array"])



numpy.ndarray

In [8]:
import librosa
def load_wav_to_tensor(file_path):
    # Load the WAV file using librosa
    waveform, sample_rate = librosa.load(file_path, sr=None, mono=True)

    # Convert the waveform to a torch tensor
    tensor_waveform = waveform #torch.tensor(waveform).unsqueeze(0)

    return tensor_waveform, sample_rate

# Example usage
file_path = r"C:\work\local4test\sampleAudio\shortTestRecording.wav"
audio, sample_rate = load_wav_to_tensor(file_path)

print(type(audio), audio.shape, sample_rate)

<class 'numpy.ndarray'> (262640,) 16000


In [14]:
%%time
# load model and processor

#model_name = "openai/whisper-medium"
model_name = "openai/whisper-tiny"
processor = WhisperProcessor.from_pretrained(model_name)
model = WhisperForConditionalGeneration.from_pretrained(model_name)

CPU times: total: 922 ms
Wall time: 38 s


In [15]:
%%time

input_features = processor(audio, sampling_rate=16000, return_tensors="pt").input_features 

# generate token ids
predicted_ids = model.generate(input_features)
# decode token ids to text
transcription = processor.batch_decode(predicted_ids, skip_special_tokens=False)
print(transcription)

['<|startoftranscript|><|en|><|transcribe|><|notimestamps|> Testing 16kHz 16bts creating a PCM to see what the recognition gives us a short message for testing.<|endoftext|>']
CPU times: total: 1.55 s
Wall time: 709 ms


# model_name = "openai/whisper-medium"

audio 16sec C:\work\local4test\sampleAudio\shortTestRecording.wav

['<|startoftranscript|><|en|><|transcribe|><|notimestamps|> Testing 16 kilohertz, 16 bits, creating a PCM to see what the recognition gives us. A short message for testing.<|endoftext|>']
CPU times: total: 32.3 s
Wall time: 12.2

#model_name = "openai/whisper-tiny"
['<|startoftranscript|><|en|><|transcribe|><|notimestamps|> Testing 16kHz 16bts creating a PCM to see what the recognition gives us a short message for testing.<|endoftext|>']
CPU times: total: 1.55 s
Wall time: 709 ms
 s